src/main/kotlin/org/rust/ide/intentions/ExtractEnumVariantIntention.kt

```diff
@@ -0,0 +1,210 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.intentions
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiElement
+import com.intellij.psi.search.searches.ReferencesSearch
+import org.rust.ide.refactoring.RsInPlaceVariableIntroducer
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsVisibility
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.parentEnum
+import org.rust.lang.core.psi.ext.visibility
+import org.rust.lang.core.types.infer.TypeVisitor
+import org.rust.lang.core.types.infer.needToSubstitute
+import org.rust.lang.core.types.regions.ReEarlyBound
+import org.rust.lang.core.types.regions.Region
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.ty.TyReference
+import org.rust.lang.core.types.ty.TyTypeParameter
+import org.rust.lang.core.types.type
+
+class ExtractEnumVariantIntention : RsElementBaseIntentionAction<ExtractEnumVariantIntention.Context>() {
+    override fun getText() = "Extract enum variant"
+    override fun getFamilyName() = text
+
+    class Context(val variant: RsEnumVariant)
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val enum = element.ancestorStrict<RsEnumVariant>() ?: return null
+        if (enum.tupleFields == null && enum.blockFields == null) {
+            return null
+        }
+        if (enum.variantDiscriminant != null) {
+            return null
+        }
+        return Context(enum)
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val factory = RsPsiFactory(project)
+        val element = createElement(ctx.variant, factory)
+
+        val enum = ctx.variant.parentEnum
+        val typeParameters = filterTypeParameters(element.typeReferences, enum.typeParameterList)
+        val whereClause = enum.whereClause
+        val name = ctx.variant.name ?: return
+
+        val struct = element.createStruct(name, typeParameters, whereClause?.text ?: "")
```

Comment:
I'm not sure how to do that. If I do something like this:
```
factory.createWhereClause(
      enum.typeParameterList.lifetimeParameterList,
      enum.typeParameterList.typeParameterList
)
```
then it just generates empty `where`, since the type parameters have `typeParamBounds == null`.

Reply:
You can traverse the existing `typeParameterList` and `whereClause` to collect needed parameters with bounds.

Before:
```rust
#![feature(const_generics)]

fn foo<'a1: 'a2 + 'a3, 'a2, 'a3, T1: 'a1, T2: 'a2>() where &'a1 T1: Sync, &'a2 T2: Sync {
    /*caret*/bar::<T1>();
}

fn bar<'a1, 'a2, T>() {}
```

After:
```rust
#![feature(const_generics)]

fn foo<'a1: 'a2 + 'a3, 'a2, 'a3, T1: 'a1, T2: 'a2>() where &'a1 T1: Sync, &'a2 T2: Sync {
    baz::<'a1, 'a2, T1>();
}

fn bar<'a1, 'a2, T>() {}

fn baz<'a1: 'a2, 'a2, T1: 'a1>() where &'a1 T1: Sync {
    bar::<T1>()
}
```

But it may be too hard so you can just ignore bounds for now.

Reply:
Added where clause generation.
For lifetimes, it puts only used lifetimes into the where clause, and keeps only the lifetime bounds that are also moved to the new struct.
```rust
enum <'a, 'b, 'c> E where 'a: 'b + 'c, 'b: 'c {
    V1 { a: &'a u32, b: &'b: u32 },
    V2(&'c u32)
}
// after intention
struct <'a, 'b> V1 where 'a: 'b { a: &'a u32, b: &'b: u32 }

enum <'a, 'b, 'c> E where 'a: 'b + 'c, 'b: 'c {
    V1(V1<'a, 'b>),
    V2(&'c u32)
}
```

`'a: 'b + 'c` was shortened to `'a: 'b` because `'c` is not used in the struct. `b: 'c` was removed.

Only used type parameters are now preserved in the where clause. However, I found a problem with this:
```rust
trait Trait<S> {
    type Item;
    fn foo() -> S;
}

enum E<A, B, C> where A: Trait<B, Item=C> {
    V1 { a: A },
    V2((B, C))
}
// it generates this:
struct /*caret*/V1<A, B, C> where A: Trait<B, Item=C> { a: A }
```
It correctly collects the types `B` and `C`, which are in the bound of `A`, but the resulting code fails to compile, because they are not used in the struct. We can either:
1) Ditch the trait bound if not all of its types are used in the resulting struct
2) Add `PhantomData` to the generated struct for the missing type parameters
3) Leave as it is and let the user correct the error by removing the bounds or adding fields

I'd vote for 3), as bounds on structs/enums are not very common and the code is already quite complex. What do you think?

Reply:
Yes, I vote for 3 too.